**0. Loading and High-level Overview of the Dataset.**

In [2]:
# import necessary modules
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import os
# from dython.nominal import associations
import warnings 

# silence warnings
warnings.filterwarnings('ignore')

# plotting settings
plt.style.use(['seaborn-paper'])
plt.rcParams['font.family'] = 'helvetica'

In [ ]:
# check working directory
os.getcwd()

In [3]:
# read the data
hse_data = pd.read_csv('train.csv')

The number of rows and columns:
    rows: 1460
    columns: 81

In [4]:
# check the shape of data
hse_data.shape


(1460, 81)

In [5]:
# count data types
print(hse_data.dtypes.value_counts())


# another way
cat_cols = hse_data.dtypes[hse_data.dtypes == 'object']
num_cols = hse_data.dtypes[hse_data.dtypes != 'object']
print("The number of categorical columns is {}.".format(len(cat_cols)))
print("The number of numercial columns is {}.".format(len(num_cols)))


object     43
int64      35
float64     3
dtype: int64
The number of categorical columns is 43.
The number of numercial columns is 38.


In [6]:
# print the column names 
hse_data.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

By understanding the number of features of either numerical/categorical types, 
we are able to better plan out the necessary visualisation required to best present each columns. 
You will see this subsequently later in Section 1.

In [7]:
# check summary statistic for columns
print(hse_data.describe(include='all')) 

                 Id   MSSubClass MSZoning  LotFrontage        LotArea Street  \
count   1460.000000  1460.000000     1460  1201.000000    1460.000000   1460   
unique          NaN          NaN        5          NaN            NaN      2   
top             NaN          NaN       RL          NaN            NaN   Pave   
freq            NaN          NaN     1151          NaN            NaN   1454   
mean     730.500000    56.897260      NaN    70.049958   10516.828082    NaN   
std      421.610009    42.300571      NaN    24.284752    9981.264932    NaN   
min        1.000000    20.000000      NaN    21.000000    1300.000000    NaN   
25%      365.750000    20.000000      NaN    59.000000    7553.500000    NaN   
50%      730.500000    50.000000      NaN    69.000000    9478.500000    NaN   
75%     1095.250000    70.000000      NaN    80.000000   11601.500000    NaN   
max     1460.000000   190.000000      NaN   313.000000  215245.000000    NaN   

       Alley LotShape LandContour Utili

In [8]:
# summary statistics of SalePrice column
hse_data['SalePrice'].describe()

count      1460.000000
mean     180921.195890
std       79442.502883
min       34900.000000
25%      129975.000000
50%      163000.000000
75%      214000.000000
max      755000.000000
Name: SalePrice, dtype: float64

In [9]:
# (Jensen) check the prevalence of missing values
# df.columns[df.isnull().sum() > 0]      # filtering for columns with null values

# what is the percentage of missing columns in the bigger picture?
missing_cols = hse_data.columns[hse_data.isnull().sum() > 0]
print(missing_cols)
print(f"\nThe total % of columns with missing values: {len(missing_cols)/hse_data.shape[1] * 100:.2f} %")

Index(['LotFrontage', 'Alley', 'MasVnrType', 'MasVnrArea', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Electrical', 'FireplaceQu', 'GarageType', 'GarageYrBlt',
       'GarageFinish', 'GarageQual', 'GarageCond', 'PoolQC', 'Fence',
       'MiscFeature'],
      dtype='object')

The total % of columns with missing values: 23.46 %


(Jensen) We generated an overview statistics on the % of columns plagued with missing values. 
However, in order to better understand your data – let's look at it in a more in-depth manner.

In [10]:
# missing percentage per columns or features over the overall data, multiplied by 100%
missing_percentage = hse_data.isnull().sum() / len(hse_data) * 100
# print(missing_percentage)

# print only those with missing values percentage > 0
print(missing_percentage[missing_percentage > 0].sort_values(ascending=False))


PoolQC          99.520548
MiscFeature     96.301370
Alley           93.767123
Fence           80.753425
FireplaceQu     47.260274
LotFrontage     17.739726
GarageType       5.547945
GarageYrBlt      5.547945
GarageFinish     5.547945
GarageQual       5.547945
GarageCond       5.547945
BsmtExposure     2.602740
BsmtFinType2     2.602740
BsmtFinType1     2.534247
BsmtCond         2.534247
BsmtQual         2.534247
MasVnrArea       0.547945
MasVnrType       0.547945
Electrical       0.068493
dtype: float64


(Jensen)
 Now, this missing_pct series provides us with a better picture on whether certain columns can still be salvaged with clever feature engineering such as appropriate Data Imputation. However, at a glance, we can already observed certain columns to be deemed unsuitable for further modelling or exploration:

   
    PoolQC – 99.5% missing
    MiscFeature – 96.3% missing
    Alley – 93.7% missing
    Fence – 80.7% missing
    FireplaceQu – 47% missing
   
    
The above columns contains severe data quality issue with high % of missing values. This meant that even if we conduct data imputation, we may not be accuarately representing the data distribution of the particular feature, resulting in possible misinterpretation or skewing of the eventual models.
 
The remaining columns are only having < 20% in missing values which is not a big issue. We can keep the columns for now, and further observe the columns during the EDA process to better make a judgement on whether to keep or drop the columns prior to modelling.


In [11]:
# drop the columns that deemed unsuitable for further modelling or exploration

# create list of columns to drop
drop_cols = ['PoolQC', 'MiscFeature', 'Alley', 'Fence', 'FireplaceQu']

# drop the cols and save dataset into ./data/interim
hse_data_interim = hse_data.drop(columns=drop_cols, axis=1)


In [12]:
hse_data_interim.shape

(1460, 76)

In [13]:
# save as interim data
hse_data_interim.to_csv("train_interim.csv", index=False)

print("The hse_data_interim shape is {} rows and {} columns.".format(hse_data_interim.shape[0], hse_data_interim.shape[1]))

The hse_data_interim shape is 1460 rows and 76 columns.


(Jensen) After running through a set of high-level data exploration, we understood that the train.csv – housing_data DataFrame, is not 100% perfect. We observed that there is a total of 23.46% columns (out of 81) highlighted to contain missing/null values. We, as a data analyst/scientist exploring this dataset should keep this in mind as we proceed further with Exploratory Data Analysis. We do not need to start cleaning any data yet but eventually, we will need find ways to prepare the housing_data DataFrame in a suitable way, in order for us to conduct appropriate machine-learning and generate certain predictive results.

Hence as such, we have answered the following question in Section 0:

    * What is the shape of your data i.e. number of rows and columns?
    * For the numerical columns, what does the distributions look like?
    * What is the name of the column to be predicted?

---------------------------------------------------------------------

**1. EDA**

After we had spent some time digging through the dataset and answered a few of the questions on hand, we have now reached the Exploratory Data Analysis (EDA) stage, where we are going to utilise various visualisation methods such as Bar Chart, Pie Chart, Histograms, Scatter Plot and Line Plot to help visualise the data. After all, human are trained to retain visual information longer than written (textual) information.

Before we begin, I hope that the course on data visualisation has provided you with the necessary skills to plot several different visualisations – where each often serves a specific purpose in communicating the insights in their own way.

The following serves as a valuable cheatsheet if you are just starting out in Data Science. It provides you with a high level overview in what each visualisation attempts to communicate and the different types available.

<!-- ![VisCheatsheet.jpg](../sup3-regression/home-data-for-ml-course/VisCheatsheet.jpg) -->

<!DOCTYPE html>
<html>
<head>
</head>
<body>
<img src="http://localhost:8888/view/ai4i/VisCheatsheet.jpg" width=400 height=400 />

</body>
</html>


In [ ]:
os.getcwd()

The following questions will be answered in the following data visualization process.
    * For the numerical columns, how many missing values are there for each column?
    * For the categorical columns, how many missing values are there for each column?
    * What visualizations can you use to highlight outliers in the data?

Before we move on, remember 5 columns were dropped due to the high % of missing values present within the columns. 
Hence, the new csv saved in as train_interim.csv must be loaded for the exploratory process.

In [14]:
hse_data_interim = pd.read_csv("train_interim.csv")
# check dataframe
hse_data_interim.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0   1          60       RL         65.0     8450   Pave      Reg         Lvl   
1   2          20       RL         80.0     9600   Pave      Reg         Lvl   
2   3          60       RL         68.0    11250   Pave      IR1         Lvl   
3   4          70       RL         60.0     9550   Pave      IR1         Lvl   
4   5          60       RL         84.0    14260   Pave      IR1         Lvl   

  Utilities LotConfig  ... EnclosedPorch 3SsnPorch ScreenPorch PoolArea  \
0    AllPub    Inside  ...             0         0           0        0   
1    AllPub       FR2  ...             0         0           0        0   
2    AllPub    Inside  ...             0         0           0        0   
3    AllPub    Corner  ...           272         0           0        0   
4    AllPub       FR2  ...             0         0           0        0   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
0       0      2    2008        WD         Normal     208500  
1       0      5    2007        WD         Normal     181500  
2       0      9    2008        WD         Normal     223500  
3       0      2    2006        WD        Abnorml     140000  
4       0     12    2008        WD         Normal     250000  

[5 rows x 76 columns]

(Jensen) Thereafter, it will be useful for us to have two lists – consisting of the cols of each respective types, numerical and categorical. This will help us when I demonstrate how to automate the process of generating appropriate visualisations to ease the process of visualising ~70+ columns.

In [15]:
# list non-numeric and numeric column names
cat_cols = hse_data_interim.dtypes[hse_data.dtypes == 'object'].index.tolist()
num_cols = hse_data_interim.dtypes[hse_data.dtypes != 'object'].index.tolist()


print("The number of categorical columns is {}.".format(len(cat_cols)))
print("The number of numercial columns is {}.".format(len(num_cols)))

The number of categorical columns is 38.
The number of numercial columns is 38.


(Jensen) Let's take a look at the numerical columns. One reasons to do a further inspection of the numerical columns is because, often than not, discrete values (e.g., 1,2,3,4 – which represents categories) are represented as numerical columns. Unlike continuous values – like SalePrice, trying to plot a chart like a Histogram on discrete may not provide as much insights as compared to something like a Boxplot.

In [16]:
# Jensen's
# hence, let's seperate the columns based on the description file found in '../../data/raw/data_description.txt'

uniq_threshold = 20  # arbitrary number
discrete_num_cols = [col for col in hse_data_interim.columns if (hse_data_interim[col].nunique() < uniq_threshold) & (hse_data_interim[col].dtype != 'object') ]

# saving the continuous cols in a seperate list for ease of usage
cont_num_cols = [col for col in num_cols if col not in discrete_num_cols]

# running a small check; total of 15 discrete columns
# print(hse_data_interim.loc[:,discrete_num_cols])
# printing summmary 
print(hse_data_interim.loc[:,discrete_num_cols].describe())

# sanity check to make sure len(disc_num_cols) + (cont_num_cols) == len(num_cols)

assert(len(discrete_num_cols) + len(cont_num_cols) == len(num_cols))

        MSSubClass  OverallQual  OverallCond  BsmtFullBath  BsmtHalfBath  \
count  1460.000000  1460.000000  1460.000000   1460.000000   1460.000000   
mean     56.897260     6.099315     5.575342      0.425342      0.057534   
std      42.300571     1.382997     1.112799      0.518911      0.238753   
min      20.000000     1.000000     1.000000      0.000000      0.000000   
25%      20.000000     5.000000     5.000000      0.000000      0.000000   
50%      50.000000     6.000000     5.000000      0.000000      0.000000   
75%      70.000000     7.000000     6.000000      1.000000      0.000000   
max     190.000000    10.000000     9.000000      3.000000      2.000000   

          FullBath     HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  \
count  1460.000000  1460.000000   1460.000000   1460.000000   1460.000000   
mean      1.565068     0.382877      2.866438      1.046575      6.517808   
std       0.550916     0.502885      0.815778      0.220338      1.625393   
min    

(Jensen) **1.1 Visualising Distributions in Numerical Columns**

So, how are we going to visualize numerical columns? One of the questions we aim to answer through EDA, is how are the numerical columns distributed? Some first thoughts that come to mind are Histograms, Scatter Plot. And guess, what? Seaborn provides just the right visualisation that we need!

As we explain previously on why we are seperating between discrete numerical vs continuous numerical data, in this portion, we will be plotting for continuous numerical data first in order to see the distribution between the features vs the target (a bit on the correlation here.)

In [17]:
print(len(cont_num_cols))
print(cont_num_cols)

23
['Id', 'LotFrontage', 'LotArea', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'GarageYrBlt', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'MiscVal', 'SalePrice']


In [ ]:
# Jensen's
# to prevent cramped visualisations, we will seperate and plot in groups of 4 + 1 (target) 
# to better understand both the distribution of the numerical column and 
# also the relationship (scatterplot) between the numerical columns and the target.

# temp removing the target 
cont_num_cols.remove('SalePrice')
# check to ensure num of cont cols as expected
assert(len(cont_num_cols) == 22)

for i in range(0, len(cont_num_cols), 4):
    # plotting 4 + 1 (target)
    curr_cols = cont_num_cols[i:i+4] + ['SalePrice']
    sns.pairplot(hse_data_interim.loc[:,curr_cols], diag_kind='kde', size=2, dropna=True)
    
    # saving figure
    # plt.savefig(f"../reports/figures/pairplot_{i}_{i+4}.jpg")
    plt.show()

After generating respective pairplot of continuous cols (features) against SalePrice, let's do a quick overview analysis on some patterns that we can spot and should deep-dive further in subsequent EDA procedures.

Looking at the Scatterplot and Histogram, the first thing that comes to mind is that majority of the continuous numerical data are not of Normal Distributions. This may pose a potential problem if machine learning algorithms that assume normally distributed data. However, we do have other algorithms that is non-parametric in nature – meaning that no prior assumption of the data distribution is required.

Besides looking at the distribution, we should also focus on looking out of potential correlated features with either: **1) the target – SalePrice or 2) other features**. So you may ask do we need to look out of either correlations?

It is because:

**If a feature is correlated with the target, it may provide some insights on how that particular features may affect the final prediction.** For instance, looking at GrLivingArea and SalePrice, we can see that the feature is positively correlated with the target. So what exactly is GrLivingArea? Based on the data description, we understand that the feature refers to Above grade (ground) living area square feet – which may be simplfied as how many more extra living area is available above the ground floor, which logically suggests the relationship where, the more living area, the higher the potential price.

The above provides an analytical though process on how you should leverage the visualisations to help you better understand your data. Subsequently, we will leverage on these existing information, to conduct several more in-depth statistical tests (Pearson's Correlation, Spearman's Correlation) and visualisation (Correlation Plot). This enables us to make wiser decisions on which additional columns should be removed, engineered and keep for the final modelling process.

Alright! After looking at the continous numerical data, we should now start exploring the discrete continuous data instead. As mentioned previously, discrete numerical data may be in a numerical data in nature, but rather they perform similarly like a categorical features. Hence, Scatterplot may not be the best visualisation for us to retrieve valuable insights. Instead, we will be leveraging on Boxplot to help see the Interquartile Range and also highlight any potential outliers.

In [ ]:
# function def for plotting distributions of discrete cols using box vs bar 
def plot_cat_distribution(x, y, data):
    # plt.subplots(nrows, ncols, figsize=(x, y), sharey=True:y-axis will be shared among all subplots.)
    # (ax1, ax2) using tuple unpacking for multiple Axes
    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(15, 6), sharey=True)

    # plotting bar plot vs box plot
    sns.barplot(x, y, data=data, ax=ax1, palette='rocket', ci=False)
    sns.boxplot(x, y, data=data, ax=ax2, palette='rocket')

    # giving title
    fig.suptitle(f'Distribution of {x}')
    plt.tight_layout(rect=[0, 0.03, 1, 0.95])
    plt.show()

In [ ]:
# plotting boxplot for each discrete columns against the sale price

for col in discrete_num_cols:
    plot_cat_distribution(col, 'SalePrice', data=hse_data_interim)

<!DOCTYPE html>
<html>
<head>
<title></title>
</head>
<body>

<p>
    Looking at the boxplots, don't you find it easy to spot potential outliers? That is one observation, I have on my end. Looking through all the discrete features, we observed that several (or majority) of the features consists of potential outliers. Why do I say potential? Because such outliers may still be a valuable information that is part of our dataset. In essence, if you cannot justify why you would want to remove the outliers – you are basically conforming your model to assume less variability that is actually occuring in the actual "real" world dataset. Hence, a <a href="https://statisticsbyjim.com/basics/remove-outliers/">guideline</a>  which I find useful would be the following, in determine outliers:
   
   
</p>
    <ol>
      <li>A measurement error or data entry error, correct the error if possible. If you can't fix it, remove the observation because you know it's incorrect.</li>
      <li>Not a part of the population you are studying (i.e., unusual properties or conditions), you can legitimately remove the outlier.
      </li>
      <li>If the data point is a natural part of the population you are studying, you should NOT remove it.
      </li>
    </ol>
   
    
<p> 
        However, another approach – through experimental design is to run the model through two different datasets: 1) one with outliers removed, 2) one without outliers removed. These may be useful to validate your doubts on whether to remove the outliers after actually seeing the validation metrics.
</p>
    
<p> 
        Hence, for now, we will still leave the outliers in, but let's still run a bunch of summary statistics to better determine if the outliers are actual errors as opposed to just naturally occuring outliers.
</p> 
    
</body>
</html>


**1.1.1 Visualising Correlations in Numerical Columns**

<!DOCTYPE html>
<html>
<head>
<title></title>
</head>
<body>
    <div>
    <p> 
    After plotting a few charts in an attempt to understand the underlying distributions within the numerical columns, we have another question. We have tried to explain briefly earlier on, how through the pairplots, we are able to observe both the distributions as well as correlations through the scatterplots. However, as we have quite a lot of continuous numerical columns, we are unable to see the bigger pictures and identify even more potentially correlated features.
    </p> 
    <p>
       The reason why these correlated features matters, is that correlated features in general <a href="https://datascience.stackexchange.com/questions/24452/in-supervised-learning-why-is-it-bad-to-have-correlated-features">do not improve models</a>. Let's put things into perspective, given that we asked a user about the number of child he/she has, and we also asked he/she the number of son & daughters seperately. This presents in a correlation. Why? The number of child is positively correlated with number of son and number of daughter. If an user say that she has 3 daughters, the number of child would likely be 3 or more. Hence, a positive increment in both <code>sons</code> and <code>daughters</code> will result in an increase in <code>child</code> as well.
    </p>
    <p>
            In Machine Learning, there's a principle of parsimony, or <a href="https://en.wikipedia.org/wiki/Occam%27s_razor">Occam's razor</a>, where a simpler model is preferred as opposed to building a model with 100s or 1000s or features, especially when a simple model can provide similar or even better results (potential overfitting if you have so many features, resulting in poor performance).
    </p>
    <p>
            Hence, let's leverage on Seaborn's heatmap plots to better help us visually identify such correlated features, which we will then go through the data description to see why they might be correlated.
    </p>
    </div>   
</body>
</html>

# finding plot a better visualization to better observe which features 
# correlates the most with `SalePrice`

def plot_corr_map(cont_cols, figsize=(25, 20)):
    # references: https://medium.com/analytics-vidhya/kaggle-house-prices-prediction-with-linear-regression-and-gradient-boosting-c5694d9c6df4
    # subset data
    cont_data = hse_data_interim.loc[:, cont_num_cols + ['SalePrice']] 
    
    f, ax = plt.subplots(figsize=(12, 9))    
    sns.heatmap(cont_data, vmax=0.8, square=True, annot=True)


In [ ]:
# Jensen's
# plotting correlation heatmap

def plot_corr_map(cont_cols, figsize=(25,20)):
    # references: https://medium.com/@szabo.bibor/how-to-create-a-seaborn-correlation-heatmap-in-python-834c0686b88e

    # dataframe subset
    cont_data = hse_data_interim.loc[:, cont_cols+['SalePrice']]

    # define mask to isolate upper right triangle
    mask = np.triu(np.ones_like(cont_data.corr(), dtype=np.bool))

    # setting plot size
    plt.figure(figsize=figsize)
    heatmap = sns.heatmap(cont_data.corr(), mask=mask, vmin=-1, vmax=1, annot=True, cmap='BrBG')


In [ ]:
plot_corr_map(cont_num_cols)


In [ ]:
# finding plot a better visualization to better observe which features correlates the most with `SalePrice`

def plot_correlation_bar(cont_cols, figsize=(10, 15)):
    
    # subset data
    corr_sales = hse_data_interim.loc[:, cont_num_cols + ['SalePrice']].corr()[['SalePrice']]

    # setting plotting figure
    plt.figure(figsize=figsize)
    heatmap = sns.heatmap(corr_sales.sort_values(by='SalePrice', ascending=False), vmin=-1, vmax=1, annot=True, cmap='coolwarm')
    plt.show()
    

In [ ]:
plot_correlation_bar(cont_num_cols)


<!DOCTYPE html>
<html>
<head>
</head>
    <body>
        <p>
            Looking at the correlaion heatmap, we can sieve through some columns that we possibly drop in the later part of the project during the <code>Feature Selection</code> stage. Here's are some of the columns and the justifications on why these columns are highlighted:
        </p> 

            <ol>
                 <li><em>GarageYrBlt</em> – it has correlation coefficient of <code>0.83</code> with <em>YearBuilt</em>, which is logical considering that the garage either comes along with the house or built later on. Usually one would tend to observe which year the house was constructed as oppose to the garage as the garage is just small extension of the house.
                  </li>
                  <li><em>TotalBsmtSF</em> – it has a correlation coefficient of <code>0.82</code> with <em>1stFlrSF</em>, which is probable considering basement under the first floor of the house tends to be similarly sized in square feet or smaller.
                  </li>
            </ol>
          
          
     
        <p>Thus far we set our threshold at <code>>0.70</code>, as besides manually selecting features to drop, we can also depend on other feature selections method that we will be discussing in the later part of the project.<br>
            Lastly, if you have not noticed, there are other column(s) that need to be dropped not because of possible collinearity but rather the columns provides no values to the modelling later. Take a look at <code>Id</code>, it is a column to identify each record but does not mean anything to the <code>SalePrice</code>.
        </p>
        <p> 
    Hence, based on this section of the exploration, we have identified TotalBsmBlt, GarageYrBlt and Id that would be removed at the end of this section before further preprocessing.

        </p>
    </body>
</html>

**1.2 Visualising Distributions in Categorical Columns**


After attempting to plot some visualisations to understand the distributions of the numerical columns, we now turn our eyes to plotting the categorical columns. Plotting categorical columns is just similar to plotting discrete numerical columns as both have discrete values. For instance <code>0/1</code> in discrete numerical column can also be represented as <code>Yes/No</code> in a categorical column.

Beside plotting a Bar Plot, we can also use a Box Plot similar to what we did for the discrete numerical columns. This enables us to observe the 25th, 50th (median), 75th percentile of the distribution in the <code>SalePrice</code>.

So, let's first plot some bar and box plot to help understand the value counts within each columns and identify if there's any values of any columns that is either in overwhelming majority or underwhelming minority.


In [ ]:
for col in cat_cols:
    plot_cat_distribution(col, 'SalePrice', data=hse_data_interim)

As an overview, the distribution of the categorical columns provides us a good look at how the each category represents themselves in terms of both the *average* <code>SalePrice</code> and also the range of <code>SalePrice</code> it encompasses. Similarly to what we observed for the *discrete numerical* columns, we still observe possible outliers here and there, but as we are cannot be completely sure that outliers occurs non-naturally (e.g., human errors), we shall not remove these information. Still, we can conduct possible experiment subsequently to compare the results between removal and non-removal.

**1.2.1 Visualising Correlations in Categorical Columns**

Many of you may have understood the concept of Pearson R coefficient used in determining correlation between two continous fields. However, we can also use something known as <a href="https://en.wikipedia.org/wiki/Cram%C3%A9r%27s_V">Cramer's V</a> to help us to understand the relationship between two categorical fields.

In [ ]:
# ROC, AUC,


_, ax = plt.subplots(figsize=(25,25))
associations(hse_data_interim.loc[:,cat_cols], ax=ax)
plt.show()

In [ ]:
top_correlations = hse_data_interim.corr()
top_feature_columns = top_correlations['SalePrice'][top_correlations['SalePrice'].values > 0.2].index.values
top_feature_columns

In [ ]:
# references: https://medium.com/analytics-vidhya/kaggle-house-prices-prediction-with-linear-regression-and-gradient-boosting-c5694d9c6df4

# Handling Missing Values for 19 features which have missing values mentioned above
hse_data_interim['GarageYrBlt'] = hse_data_interim['GarageYrBlt'].fillna(0)
# filling in missing GarageYrBuilt values with zeros.  
# But this may not be the most logical approach - refer to this discussion below for mor perspective
# https://www.kaggle.com/c/house-prices-advanced-regression-techniques/discussion/60143

# similary fillingup na valuse for couple of other features
hse_data_interim['LotFrontage'] = hse_data_interim.groupby('Neighborhood')['LotFrontage'].transform(lambda x: x.fillna(x.median()))
hse_data_interim['MasVnrArea'] = hse_data_interim['MasVnrArea'].fillna(0)
heat_map_with_top_correlated_features = np.append(top_feature_columns[-12:], np.array(['SalePrice']))
pearson_correlation_coefficients = np.corrcoef(hse_data_interim[heat_map_with_top_correlated_features[::-1]].T)
plt.figure(figsize=(16,16))
sns.set(font_scale=1)
with sns.axes_style('white'):
    sns.heatmap(pearson_correlation_coefficients, yticklabels=heat_map_with_top_correlated_features[::-1], xticklabels=heat_map_with_top_correlated_features[::-1], fmt='.2f', annot_kws={'size': 10}, annot=True, square=True, cmap=None)

Now lets glance over the above heatmap to understand some points just visualy looking at the correlation number

‘OverallQual’, ‘GrLivArea’ and ‘TotalBsmtSF’ have strong correlation with ‘SalePrice’.

While ‘GarageCars’ and ‘GarageArea’ also have strong correlation, but they are mostly mutually dependent i.e. they are NOT linearly independent of each other, i.e. there is a high-multicollinearity (0.88 as we can see in the figure) between them. Because, the number of cars that fit into the garage is dependent of the garage area. Hence, we just need one of these variables in our analysis (we can decide to keep ‘GarageCars’ as its correlation with ‘SalePrice’ is higher).

Same mutual dependence applies to the two features ‘TotalBsmtSF’ and ‘1stFloor’ . We will take only ‘TotalBsmtSF’ in our feature-engineering.

AND ALSO ‘TotRmsAbvGrd’ and ‘GrLivArea’, and we will only take ‘GrLivArea’

# references: Jensen's: https://nbviewer.jupyter.org/github/wtlow003/aiig-suss/blob/main/learning_materials/regression/notebooks/01-lwt-initial-exploratory-data-analysis.ipynb

Likewise to using Pearson's R for correlations between continuous features, we can also use Cramer's V and Theil's U in a similar fashion. The heatmap that we have plotted above is symmetrical in nature - which is one of the properties of Cramer's V. It comes with certain pitfalls, but in this first EDA, we will not discuss about it. Here are some features that we can possibly drop:

1. *GarageQual* – it has a value of <code>0.7</code> with *GarageCond*, which higlights a high association between the two nominal features. This association makes sense considering that both *GarageQual* and GarageCond are trying to quantify the garage in a measurable way. Hence, we can just consider keeping 1 of the columns instead.

2. *Exterior2nd* – it has a value of <code>0.76</code> with *Exterior1st*, which highlights a high assocation between the two nominal features. Considering that both describes the exterior covering on the house, where Exterior2nd is applicable, if more than one material is used. Instead of dropping this feature, we can consider conducting <code>Feature Engineering</code> and generate another column on how many exterior finishing is used, while we still use the main exterior finishing – Exterior1st as a feature.

**2. Concluding Remarks**

So after plotting and spending some time looking through the plot and generating relevant hypothesis on which features we can possible drop to reduce the features required for the model, and also answer some questions we had on hand prior to conducting EDA. To bring ourselves back to the current process of the data preprocesssing – we have 4 columns to remove. They are:

1. <code>GarageYrBlt</code>
2. <code>TotalBsmtSF</code>
3. <code>Id</code>
4. <code>GarageQual</code>

We will be saving this DataFrame in its interim stage while we proceed on with further <code>Feature Selection & Feature Engineering</code>.

In [ ]:
# columns to drop
drop_cols = ['GarageYrBlt', 'TotalBsmtSF', 'Id', 'GarageQual']

hse_data_interim = hse_data_interim.drop(drop_cols, axis=1)


In [ ]:
hse_data_interim.to_csv("train_csv_EDA.csv", index=False) 

# check the current updated dataframe shape to ensure only 4 columns are dropped
print(f"The current DataFrame shape is: {hse_data_interim.shape[0]} rows, {hse_data_interim.shape[1]} columns.")

**--------------My own random Viz------------------**

In [ ]:
Distribution for numerical features
- many data points
- scatter plot 
- box plot for summary stats

In [ ]:
sns.scatterplot(x='YearBuilt', y='SalePrice', data=df, hue='SaleCondition')
plt.show()

In [ ]:
# try: plotting as FacetGrid: can create subplots
sns.set_context('notebook')
sns.set_palette('RdBu')
sns.set_style('darkgrid')
g = sns.catplot(x='YearBuilt', y='SalePrice', data=df, kind='box', col="SaleCondition")

# give a title and position the title
g.fig.suptitle("Year Built vs. Sale Price", y=1.03)

# give titles for each column
g.set_titles("this is column name: {col_name}")
plt.show()

In [ ]:
# try: plotting as AxesSubplots: only creates single plot; scatterplot(), countplot(), etc.
sns.set_context('paper')
sns.set_palette('RdBu')
sns.set_style('darkgrid')
g = sns.boxplot(x='YearBuilt', y='SalePrice', data=df)

# give a title and position the title
g.set_title("Year Built vs. Sale Price", y=1.03)


plt.show()

In [ ]:
Target variable (column to be predicted)  --> SalePrice

In [ ]:
How are the various attributes correlated to the outcome variable?
yearBuilt, lotArea, miscval

In [ ]:
# For the numerical columns, how many missing values are there for each column? 
# find missing values for each column
missing_val = df.isnull().sum()
print(missing_val)

In [ ]:
df[df['MiscVal'].isnull()]

In [ ]:
df[df['MiscFeature'].isnull()]

In [ ]:
For the categorical columns, how many missing values are there for each column?

In [ ]:
Visualize to highlight the outliers

In [ ]:
df.info()


In [ ]:
df.head(15)
# 

In [ ]:
numeric_feats = df.select_dtypes(include=['float64', 'int64'])
numeric_feats

In [ ]:
object_feats = df.select_dtypes(include=['object'])
object_feats

In [ ]:
df['YearBuilt']